File: backend/src/pipeline/normalize.ts
import { prisma } from "../db/prisma.js";
import { canonicalizeUrl } from "../lib/url.js";
import { contentHash, tokenSetString } from "../lib/text.js";
import { logger } from "../lib/logger.js";

const log = logger("normalize");

/**
 * Turn a RawItem into a normalized Article, applying exact dedupe.
 * Returns the new articleId, or null if it was an exact duplicate.
 */
export async function normalizeRawItem(rawItemId: string): Promise<string | null> {
  const raw = await prisma.rawItem.findUnique({ where: { id: rawItemId } });
  if (!raw) return null;
  if (raw.status === "PARSED" || raw.status === "DUPLICATE") {
    const existing = await prisma.article.findUnique({ where: { rawItemId } });
    return existing?.id ?? null;
  }

  const title = (raw.rawTitle ?? "").trim();
  const body = (raw.rawContent ?? "").trim();
  if (!title) {
    await prisma.rawItem.update({ where: { id: rawItemId }, data: { status: "FAILED" } });
    return null;
  }

  const canonicalUrl = canonicalizeUrl(raw.rawUrl);
  const hash = contentHash(title, body);

  // Exact dedupe: same content hash or same canonical URL already normalized.
  const dup = await prisma.article.findFirst({
    where: {
      OR: [
        { contentHash: hash },
        ...(canonicalUrl ? [{ canonicalUrl }] : []),
      ],
    },
    select: { id: true },
  });
  if (dup) {
    await prisma.rawItem.update({ where: { id: rawItemId }, data: { status: "DUPLICATE" } });
    log.debug(`raw ${rawItemId} is exact duplicate of article ${dup.id}`);
    return null;
  }

  const article = await prisma.article.create({
    data: {
      rawItemId: raw.id,
      sourceId: raw.sourceId,
      canonicalUrl,
      title,
      body,
      summary: body.slice(0, 280) || null,
      publishedAt: raw.publishedAt,
      contentHash: hash,
      tokenSet: tokenSetString(`${title} ${body}`),
    },
  });
  await prisma.rawItem.update({ where: { id: rawItemId }, data: { status: "PARSED" } });
  return article.id;
}

File: backend/internal/pipeline/normalize.go
// Package pipeline ports the WorldSignal processing stages (fetch → normalize →
// cluster → enrich → match → send) from backend/src/pipeline.
package pipeline

import (
	"context"
	"strings"

	"github.com/worldsignal/backend/internal/db"
	"github.com/worldsignal/backend/internal/textutil"
	"github.com/worldsignal/backend/internal/urlutil"
)

// NormalizeRawItem turns a RawItem into a normalized Article, applying exact
// dedupe. Returns the new article id, or "" if it was a duplicate/failed.
// Mirrors normalizeRawItem in normalize.ts.
func NormalizeRawItem(ctx context.Context, d *db.DB, rawItemID string) (string, error) {
	raw, err := d.GetRawItem(ctx, rawItemID)
	if err != nil || raw == nil {
		return "", err
	}
	if raw.Status == "PARSED" || raw.Status == "DUPLICATE" {
		return d.ArticleIDByRawItem(ctx, rawItemID)
	}

	title := strings.TrimSpace(deref(raw.RawTitle))
	body := strings.TrimSpace(deref(raw.RawContent))
	if title == "" {
		return "", d.SetRawItemStatus(ctx, rawItemID, "FAILED")
	}

	var canonical *string
	if c, ok := urlutil.Canonicalize(deref(raw.RawURL)); ok {
		canonical = &c
	}
	hash := textutil.ContentHash(title, body)

	dup, err := d.FindDuplicateArticle(ctx, hash, canonical)
	if err != nil {
		return "", err
	}
	if dup != "" {
		return "", d.SetRawItemStatus(ctx, rawItemID, "DUPLICATE")
	}

	var bodyPtr, summaryPtr *string
	if body != "" {
		bodyPtr = &body
		sum := sliceRunes(body, 280)
		summaryPtr = &sum
	}

	id, err := d.CreateArticle(ctx, db.NewArticle{
		RawItemID: raw.ID, SourceID: raw.SourceID, CanonicalURL: canonical,
		Title: title, Body: bodyPtr, Summary: summaryPtr, PublishedAt: raw.PublishedAt,
		ContentHash: hash, TokenSet: textutil.TokenSetString(title + " " + body),
	})
	if err != nil {
		return "", err
	}
	if err := d.SetRawItemStatus(ctx, rawItemID, "PARSED"); err != nil {
		return "", err
	}
	return id, nil
}

func deref(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}

// sliceRunes returns the first n runes (approximating JS String.slice for BMP).
func sliceRunes(s string, n int) string {
	r := []rune(s)
	if len(r) > n {
		r = r[:n]
	}
	return string(r)
}
